### ISSUE TITLE
N806 (non-lowercase-variable-in-function) for local variable in function

### ISSUE THREAD

**Issue by opk12** [astral-sh/ruff#13550]
mypy supports `Final` for local variables - it errors out, if `DIR` is assigned twice.

Constants should be uppercase, but ruff gives `N806` [non-lowercase-variable-in-function](https://docs.astral.sh/ruff/rules/non-lowercase-variable-in-function).

```
import pathlib
from typing import Final


def require_files() -> None:
    DIR: Final = pathlib.Path("/...")

    if not (DIR / "file_1").exists():
        print("Missing file 1")

    if not (DIR / "file_2").exists():
        print("Missing file 2")
```

```
$ ruff  check a.py  --no-cache  --select ALL  --preview --isolated
a.py:6:5: N806 Variable `DIR` in function should be lowercase
```

**Comment by AlexWaygood** [comment]
Thanks for the issue! Hmm, `DIR` here isn't actually a constant, though, right? Since it's a local variable, you'll be creating a new `DIR` variable each time you call the function. I think you could make both Ruff and mypy happy here by moving it out of the function into the global scope (and your code would probably run slightly faster too!).

**Comment by AlexWaygood** [comment]
Sure. But if it's a local variable, then by definition it's not a constant, since you're creating it fresh every time the function is called. It's just a `Final` local variable -- so I think Ruff's complaint is correct here :-)

Only global constants that are created once should be uppercase, not local variables that are created every time the function is called. `Final` indicates that a variable is not mutated after being initialised -- this can be thought of as "necessary but not sufficient" for a variable to satisfy the definition of a global constant.

The naming convention here derives from [the relevant section in PEP 8](https://peps.python.org/pep-0008/#constants), which says:

> Constants are usually defined on a module level and written in all capital letters with underscores separating words. Examples include `MAX_OVERFLOW` and `TOTAL`.

In other words, I think either of these would satisfy both mypy and Ruff:

```py
import pathlib
from typing import Final


DIR: Final = pathlib.Path("/...")


def require_files() -> None:
    if not (DIR / "file_1").exists():
        print("Missing file 1")

    if not (DIR / "file_2").exists():
        print("Missing file 2")
```

or this:

```py
import pathlib
from typing import Final


def require_files() -> None:
	dir: Final = pathlib.Path("/...")

    if not (dir / "file_1").exists():
        print("Missing file 1")

    if not (dir / "file_2").exists():
        print("Missing file 2")
```

**Comment by opk12** [comment]
> `Final` indicates that a variable is not mutated after being initialised -- this can be thought of as "necessary but not sufficient" for a variable to satisfy the definition of a global constant.

Uhm, but people say that Python constants is just the convention of not reassigning to upper-case vars? I think the upper case is intended to look weird on the left-hand side of an assignment, not intended to mark expensive initialization.

But if ruff aligns to mypy, it becomes possible to warn about a local constant with an expensive initialization. I don't follow why your answer is so much focused on how often the variable is initialized, this ticket is not about performance - this ticket is about having to choose between a global `_MINUTES = 10` (ugly name, pollutes module namespace, puts related code apart) and a local `minutes: Final = 10` (fits the Python concept of constant, so should be upper-case).

PEP 8 says `Constants are` *usually* `defined on a module level`. For comparison, the other major programming languages do allow local constants.

A variable is not recreated, nor at the Python language level, nor at runtime. A function call has its own independent variables and call stack.

As said, mypy is already fine with the code.

**Comment by opk12** [comment]
> The principle of data hiding tells us to limit exposure of details. Constants are details. The wider the scope the wider the exposure. The correct scope for variables and constants is the narrowest scope that works. Having them visible to code that doesn’t need them harms readability.

https://softwareengineering.stackexchange.com/a/422719

**Comment by opk12** [comment]
Yes, on a second thought, if distinguishing vars from constants is necessary for readability, it likely means the function is doing too many things. Thank you very much for your answers.
